message/pool/message.go

```diff
@@ -387,3 +468,54 @@ func (r *Message) ReadBody() ([]byte, error) {
 	}
 	return payload[:n], nil
 }
+
+func (r *Message) toMessage() (message.Message, error) {
+	payload, err := r.ReadBody()
+	if err != nil {
+		return message.Message{}, err
+	}
+	m := r.msg
+	m.Payload = payload
+	return m, nil
+}
+
+func (r *Message) MarshalWithEncoder(encoder Encoder) ([]byte, error) {
+	msg, err := r.toMessage()
+	if err != nil {
+		return nil, err
+	}
+	size, err := encoder.Size(msg)
+	if err != nil {
+		return nil, err
+	}
+	if len(r.bufferMarshal) < size {
+		r.bufferMarshal = append(r.bufferMarshal, make([]byte, size-len(r.bufferMarshal))...)
+	}
+	n, err := encoder.Encode(msg, r.bufferMarshal)
+	if err != nil {
+		return nil, err
+	}
+	r.bufferMarshal = r.bufferMarshal[:n]
```

Comment:
Is this an optimization? Why not always allocate the returned array? 

Couldn't something like this happen:
```golang
data1 := r.MarshalWithEncoder(encoder1)
...
data2 := r.MarshalWithEncoder(encoder2)

...do something with data1...
```

If the buffer from the first call is large enough for the second call, then data1 will be corrupted.